### PR TITLE
Add from_iso8601_* and to_iso8601

### PR DESCRIPTION
**Pull request by haozhun**
Issue #1323 

@dain

### CODE REVIEW

**Comment by electrum** [comment]
Start commit messages with uppercase

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java [hunk 1]
```diff
@@ -159,6 +163,35 @@ public static double toUnixTimeFromTimestampWithTimeZone(@SqlType(StandardTypes.
...
+        long millisUtc = unpackMillisUtc(timestampWithTimeZone);
+        DateTimeZone dateTimeZone = unpackDateTimeZone(timestampWithTimeZone);
+        return utf8Slice(ISODateTimeFormat.dateTime().print(new DateTime(millisUtc, dateTimeZone)));
+    }
+
+    @ScalarFunction("from_iso8601")
+    @SqlType(StandardTypes.TIMESTAMP_WITH_TIME_ZONE)
+    public static long fromISO8601(ConnectorSession session, @SqlType(StandardTypes.VARCHAR) Slice iso8601DateTime)
+    {
+        System.out.println(getChronology(session.getTimeZoneKey()));
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java]: Remove this

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java [hunk 2]
```diff
@@ -159,6 +163,34 @@ public static double toUnixTimeFromTimestampWithTimeZone(@SqlType(StandardTypes.
         return unpackMillisUtc(timestampWithTimeZone) / 1000.0;
     }
 
+    @ScalarFunction("to_iso8601")
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice toISO8601(ConnectorSession session, @SqlType(StandardTypes.TIMESTAMP) long timestamp)
+    {
+        DateTimeFormatter formatter = ISODateTimeFormat.dateTime()
+                .withChronology(getChronology(session.getTimeZoneKey()));
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java]: We should talk about if this should always render in UTC or session time zone.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java [hunk 3]
```diff
@@ -159,6 +163,34 @@ public static double toUnixTimeFromTimestampWithTimeZone(@SqlType(StandardTypes.
...
+        return utf8Slice(formatter.print(new DateTime(timestamp)));
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java]: no need to create a `DateTime`, just pass the `long` directly to `print`.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java [hunk 4]
```diff
@@ -159,6 +163,34 @@ public static double toUnixTimeFromTimestampWithTimeZone(@SqlType(StandardTypes.
...
+    }
+
+    @ScalarFunction("to_iso8601")
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice toISO8601FromTimestampWithTimeZone(@SqlType(StandardTypes.TIMESTAMP_WITH_TIME_ZONE) long timestampWithTimeZone)
+    {
+        long millisUtc = unpackMillisUtc(timestampWithTimeZone);
+        DateTimeZone dateTimeZone = unpackDateTimeZone(timestampWithTimeZone);
+        return utf8Slice(ISODateTimeFormat.dateTime().print(new DateTime(millisUtc, dateTimeZone)));
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java]: Same here.  Just search the code base for other uses of `DateTimeFormatter.print`.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java (hunk 2 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java]: The outcome of the talk is session time zone.

**Comment by haozhun** [comment]
@dain, I resolved the issues mentioned in your comments and added documents / release notes

File: presto-docs/src/main/sphinx/release/release-0.97.rst [hunk 5]
```diff
@@ -0,0 +1,8 @@
+============
+Release 0.97
+============
+
+General Changes
+---------------
+
+* Add ``from_iso8601`` and ``to_iso8601`` date time functions.
```
Comment (dain) [inline presto-docs/src/main/sphinx/release/release-0.97.rst]: Use  doc function references like this:
```
:func:`from_iso8601`
```

**Comment by haozhun** [comment]
Fixed

File: presto-docs/src/main/sphinx/functions/datetime.rst [hunk 6]
```diff
@@ -74,6 +78,10 @@ Date and Time Functions
 
     This is an alias for ``current_timestamp``.
 
+.. function:: to_iso8601(timestamp) -> varchar
+
+    Returns an ISO8601 formatted string, based on the given ``timestamp`` optionally with timezone.
```
Comment (dain) [inline presto-docs/src/main/sphinx/functions/datetime.rst]: you don't need a comma here

File: presto-docs/src/main/sphinx/release/release-0.97.rst [hunk 7]
```diff
@@ -11,3 +11,4 @@ General Changes
 * Fix possible infinite loop when decompressing ORC data.
 * Fix an issue where ``NOT`` clause was being ignored in ``NOT BETWEEN`` predicates.
 * Fix a planning issue in queries that use ``SELECT *``, window functions and implicit coercions.
+* Add :func:`from_iso8601` and :func:`to_iso8601` date time functions.
```
Comment (dain) [inline presto-docs/src/main/sphinx/release/release-0.97.rst]: You'll need to move this to the 0.98 file

**Comment by dain** [comment]
Looks good

File: presto-docs/src/main/sphinx/functions/datetime.rst [hunk 8]
```diff
@@ -53,6 +53,10 @@ Date and Time Functions
     Returns the current time zone in the format defined by IANA
     (e.g., ``America/Los_Angeles``) or as fixed offset from UTC (e.g., ``+08:35``)
 
+.. function:: from_iso8601(s) -> timestamp with time zone
+
+    Returns a timestamp with time zone parsed from the ISO8601 formatted string ``s``
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/datetime.rst]: This should end in a period.

File: presto-docs/src/main/sphinx/functions/datetime.rst [hunk 9]
```diff
@@ -53,6 +53,10 @@ Date and Time Functions
     Returns the current time zone in the format defined by IANA
     (e.g., ``America/Los_Angeles``) or as fixed offset from UTC (e.g., ``+08:35``)
 
+.. function:: from_iso8601(s) -> timestamp with time zone
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/datetime.rst]: Name the variable `string` not `s`. See other function documentation for examples.

File: presto-docs/src/main/sphinx/functions/datetime.rst (hunk 6 above)
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/datetime.rst]: This should be `ISO 8601`

File: presto-docs/src/main/sphinx/release/release-0.97.rst (hunk 7 above)
Comment (electrum) [inline presto-docs/src/main/sphinx/release/release-0.97.rst]: Remove "date time"

File: presto-docs/src/main/sphinx/functions/datetime.rst (hunk 6 above)
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/datetime.rst]: Formats ``timestamp`` as an ISO 8601 string.

File: presto-docs/src/main/sphinx/functions/datetime.rst (hunk 8 above)
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/datetime.rst]: "ISO 8601"

Comment (electrum) [inline presto-docs/src/main/sphinx/functions/datetime.rst]: Parses the ISO 8601 formatted ``string`` into a ``timestamp with time zone``.

**Comment by electrum** [comment]
@martint wanted to name these differently.

**Comment by haozhun** [comment]
Rebased, updated, and added documentations.

File: presto-docs/src/main/sphinx/release/release-0.104.rst [hunk 10]
```diff
@@ -7,3 +7,5 @@ General Changes
 
 * Handle thread interruption in StatementClient.
 * Fix CLI hang when server becomes unreachable during a query.
+* Add :func:`from_iso8601_timestamp`, :func:`from_iso8601_date`, and
```
Comment (martint) [inline presto-docs/src/main/sphinx/release/release-0.104.rst]: Move this to the latest release notes

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestDateTimeFunctions.java [hunk 11]
```diff
@@ -154,6 +157,22 @@ public void testToUnixTime()
     }
 
     @Test
+    public void testFromISO8601()
+    {
+        assertFunction("from_iso8601_timestamp('" + TIMESTAMP.toString(ISODateTimeFormat.dateTime()) + "')", TIMESTAMP_WITH_TIME_ZONE, toTimestampWithTimeZone(TIMESTAMP_WITH_NUMERICAL_ZONE));
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestDateTimeFunctions.java]: The argument to the function should be a constant. Relying on the output of TIMESTAMP.toString() might mask issues if that call is sensitive to the VM timezone, etc.

**Comment by martint** [comment]
Other than those minor comments, it looks good

**Pull request merged**